site/src/Librecores/ProjectRepoBundle/Controller/ProjectController.php

```diff
@@ -250,14 +258,60 @@ public function settingsAction(Request $request, $parentName, $projectName)
         $form->handleRequest($request);
 
         if ($form->isSubmitted() && $form->isValid()) {
+            $projClassification = $request->request->get('classification');
+            $deleteClassification = $request->request->get('deleteClassification');
             $em = $this->getDoctrine()->getManager();
+            // Insert classifications
+            if (isset($projClassification)) {
+                foreach ($projClassification as $classification) {
+                    $projectClassification = new ProjectClassification();
+                    $projectClassification->setProject($p);
+                    $projectClassification->setClassification($classification);
```

Comment:
$classification is not validated in any way and could be used to insert arbitrary classifications into the DB. Please add validation. You do not need to use the Symfony validation component, simply bailing out with an Exception if invalid categories are found is sufficient (to reach that state the user must have modified the HTTP request, hence we don't need to show a nice UI).

Reply:
The backend must check if the values passed to it by the frontend are allowed according to our rules. In this case, that means a classification must be valid according to ClassificationHierarchy. You can (should, and already do) *additionally* ensure in the frontend that a user can only enter valid data. But frontend-only validation is not enough: malicious actors can craft a HTTP request with arbitrary data, and use this to insert classifications into the DB which would not have been allowed in the frontend.

Reply:
So I think there will be two rules 
- `:: ` should separate the categories
- and the categories should be in the classification hierarchy.
If you have any other rules then let me know

Reply:
Almost there: the categories should not only be in the classification hierarchy, but the full category should be valid according to the hierarchy.

Reply:
Got it, Thanks updating the PR soon